src/write.rs

```diff
@@ -450,6 +465,62 @@ mod tests {
         assert_eq!(w.oplog.len(), 2);
     }
 
+    #[test]
+    fn wait_test() {
+        use std::sync::{Arc, Barrier};
+        use std::thread;
+        let (mut w, _r) = crate::new::<i32, _>();
+
+        // Case 1: If epoch is set to default.
+        let test_epochs: crate::Epochs = Default::default();
+        let mut test_epochs = test_epochs.lock().unwrap();
+        // since there is no epoch to waiting for, wait function will return immediately.
+        w.wait(&mut test_epochs);
+
+        // Case 2: If one of the reader is still reading(epoch is odd and count is same as in last_epoch)
+        // and wait has been called.
+        let held_epoch = Arc::new(AtomicUsize::new(1));
+
+        w.last_epochs = vec![2, 2, 1];
+        let mut epochs_slab = Slab::new();
+        epochs_slab.insert(Arc::new(AtomicUsize::new(2)));
+        epochs_slab.insert(Arc::new(AtomicUsize::new(2)));
+        epochs_slab.insert(Arc::clone(&held_epoch));
+
+        let barrier = Arc::new(Barrier::new(2));
+
+        let is_waiting = Arc::clone(&w.is_waiting);
+
+        // check writers waiting state before calling wait.
+        let is_waiting_v = is_waiting.load(Ordering::Relaxed);
+        assert_eq!(false, is_waiting_v);
+
+        let barrier2 = Arc::clone(&barrier);
+        let test_epochs = Arc::new(Mutex::new(epochs_slab));
+        let wait_handle = thread::spawn(move || {
+            barrier2.wait();
+            let mut test_epochs = test_epochs.lock().unwrap();
+            w.wait(&mut test_epochs);
+        });
+
+        barrier.wait();
+
+        // make sure that writer wait() will call first, only then allow to updates the held epoch.
+        while !is_waiting.load(Ordering::Relaxed) {
+            thread::yield_now();
+        }
+
+        held_epoch.fetch_add(1, Ordering::SeqCst);
+
+        // join to make sure that wait must return after the progress/increment
+        // of held_epoch.
+        let _ = wait_handle.join();
+
+        // check that writes waiting state is set to false after wait returns.
```

Comment:
I think I'd just remove this assertion — it doesn't really add much, and anything that's in the test that's not relevant _to_ the test might confuse later readers :)